File: android/app/src/main/kotlin/com/goalmaker/app/data/auth/SupabaseAuthGateway.kt
package com.goalmaker.app.data.auth

import com.goalmaker.app.application.auth.AuthGateway
import com.goalmaker.app.application.auth.AuthResult
import com.goalmaker.app.application.auth.AuthSession
import com.goalmaker.app.domain.account.EmailAddress
import com.goalmaker.app.domain.account.SignInCode
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.OtpType
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.auth.exception.AuthRestException
import io.github.jan.supabase.auth.providers.builtin.OTP
import io.github.jan.supabase.auth.status.SessionStatus
import io.github.jan.supabase.exceptions.HttpRequestException
import io.github.jan.supabase.exceptions.RestException
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn

/** [AuthGateway] over Supabase Auth's email one-time codes. */
class SupabaseAuthGateway(
    private val client: SupabaseClient,
    scope: CoroutineScope,
) : AuthGateway {

    override val session: StateFlow<AuthSession> = client.auth.sessionStatus
        .map(::toSession)
        .stateIn(scope, SharingStarted.Eagerly, AuthSession.Loading)

    override suspend fun sendCode(email: EmailAddress): AuthResult = attempt {
        client.auth.signInWith(OTP) {
            this.email = email.value
            createUser = true
        }
    }

    override suspend fun verifyCode(email: EmailAddress, code: SignInCode): AuthResult = attempt {
        client.auth.verifyEmailOtp(type = OtpType.Email.EMAIL, email = email.value, token = code.value)
    }

    override suspend fun signOut() {
        try {
            client.auth.signOut()
        } catch (cancelled: CancellationException) {
            throw cancelled
        } catch (_: Exception) {
            // Offline: the local session is cleared anyway; the server token expires on its own.
            client.auth.clearSession()
        }
    }

    private fun toSession(status: SessionStatus): AuthSession = when (status) {
        is SessionStatus.Initializing -> AuthSession.Loading
        is SessionStatus.Authenticated -> AuthSession.SignedIn(status.session.user?.email.orEmpty())
        // A failed refresh (usually offline) keeps the stored session; the user is still signed in.
        is SessionStatus.RefreshFailure ->
            AuthSession.SignedIn(client.auth.currentSessionOrNull()?.user?.email.orEmpty())
        is SessionStatus.NotAuthenticated -> AuthSession.SignedOut
    }

    private suspend fun attempt(block: suspend () -> Unit): AuthResult = try {
        block()
        AuthResult.Success
    } catch (cancelled: CancellationException) {
        throw cancelled
    } catch (error: AuthRestException) {
        val code = error.error.lowercase()
        when {
            "expired" in code || "invalid" in code || "otp" in code -> AuthResult.WrongOrExpiredCode
            "rate" in code || error.statusCode == TOO_MANY_REQUESTS -> AuthResult.TooManyRequests
            else -> AuthResult.Failed(error.description ?: error.error)
        }
    } catch (error: RestException) {
        if (error.statusCode == TOO_MANY_REQUESTS) AuthResult.TooManyRequests else AuthResult.Failed(error.error)
    } catch (_: HttpRequestException) {
        AuthResult.Offline
    }

    private companion object {
        const val TOO_MANY_REQUESTS = 429
    }
}

File: windows/src/GoalMaker.Infrastructure/Auth/SupabaseAuthGateway.cs
using System.Net;
using GoalMaker.Core.Account;
using GoalMaker.Core.Auth;
using Supabase.Gotrue;
using Supabase.Gotrue.Exceptions;
using Supabase.Gotrue.Interfaces;

namespace GoalMaker.Infrastructure.Auth;

/// <summary><see cref="IAuthGateway"/> over Supabase Auth's email one-time codes.</summary>
public sealed class SupabaseAuthGateway : IAuthGateway
{
    private readonly Supabase.Client client;

    public SupabaseAuthGateway(Supabase.Client client)
    {
        this.client = client;
        client.Auth.AddStateChangedListener(OnAuthStateChanged);
    }

    public event EventHandler<AuthSession>? SessionChanged;

    public AuthSession Session { get; private set; } = new AuthSession.Loading();

    public async Task InitializeAsync(CancellationToken cancellationToken)
    {
        try
        {
            await client.InitializeAsync().ConfigureAwait(false);
        }
        catch (Exception error) when (error is GotrueException or HttpRequestException)
        {
            // Offline at start-up: a stored session stays usable and refreshes later.
        }

        Publish(FromCurrentUser());
    }

    public Task<AuthResult> SendCodeAsync(EmailAddress email, CancellationToken cancellationToken) =>
        AttemptAsync(() => client.Auth.SignInWithOtp(new SignInWithPasswordlessEmailOptions(email.Value)));

    public Task<AuthResult> VerifyCodeAsync(EmailAddress email, SignInCode code, CancellationToken cancellationToken) =>
        AttemptAsync(async () =>
        {
            var session = await client.Auth.VerifyOTP(email.Value, code.Value, Constants.EmailOtpType.Email)
                .ConfigureAwait(false);
            if (session?.User is null)
            {
                throw new GotrueException("Token has expired or is invalid", FailureHint.Reason.UserBadLogin);
            }
        });

    public async Task SignOutAsync()
    {
        try
        {
            await client.Auth.SignOut().ConfigureAwait(false);
        }
        catch (Exception error) when (error is GotrueException or HttpRequestException)
        {
            // Offline: the stored session is removed anyway; the server token expires on its own.
        }

        Publish(new AuthSession.SignedOut());
    }

    private void OnAuthStateChanged(IGotrueClient<User, Session> sender, Constants.AuthState state) =>
        Publish(state == Constants.AuthState.SignedOut ? new AuthSession.SignedOut() : FromCurrentUser());

    private AuthSession FromCurrentUser() =>
        client.Auth.CurrentUser is { } user
            ? new AuthSession.SignedIn(user.Id ?? string.Empty, user.Email ?? string.Empty)
            : new AuthSession.SignedOut();

    private void Publish(AuthSession session)
    {
        if (session == Session)
        {
            return;
        }

        Session = session;
        SessionChanged?.Invoke(this, session);
    }

    private static async Task<AuthResult> AttemptAsync(Func<Task> action)
    {
        try
        {
            await action().ConfigureAwait(false);
            return new AuthResult.Success();
        }
        catch (GotrueException error)
        {
            return error.Reason switch
            {
                FailureHint.Reason.Offline or FailureHint.Reason.NetworkError => new AuthResult.Offline(),
                FailureHint.Reason.UserTooManyRequests => new AuthResult.TooManyRequests(),
                _ when error.StatusCode == (int)HttpStatusCode.TooManyRequests => new AuthResult.TooManyRequests(),
                _ when error.StatusCode == (int)HttpStatusCode.Forbidden
                    || error.Message.Contains("expired", StringComparison.OrdinalIgnoreCase)
                    || error.Message.Contains("invalid", StringComparison.OrdinalIgnoreCase) => new AuthResult.WrongOrExpiredCode(),
                _ => new AuthResult.Failed(error.Message),
            };
        }
        catch (HttpRequestException)
        {
            return new AuthResult.Offline();
        }
    }
}
